File: src/parsing/docblock_parser.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Sedlexing = Flow_sedlexing
module Ast = Flow_ast

(* Avoid lexing unbounded in perverse cases *)
let docblock_max_tokens = 10

type docblock_error_kind =
  | MultipleFlowAttributes
  | InvalidFlowMode of string
  | MultipleJSXAttributes
  | InvalidJSXAttribute of string option
  | MultipleJSXRuntimeAttributes
  | InvalidJSXRuntimeAttribute
  | InvalidSupportsPlatform of string
  | DisallowedSupportsPlatform

type docblock_error = Loc.t * docblock_error_kind

let attributes_rx = Str.regexp "[ \t\r\n\\*/]+"

let lines_rx = Str.regexp "\\(\r\n\\|\n\\|\r\\)"

let pragma_rx = Str.regexp "^@"

let extract_docblock =
  Docblock.(
    (* walks a list of words, returns a list of errors and the extracted info.
       if @flow or @providesModule is found more than once, the first one is used
       and an error is returned. *)
    let rec parse_attributes ~file_options ~filename (errors, info) = function
      | (loc, "@flow") :: (_, "strict") :: xs ->
        let acc =
          if info.flow <> None then
            ((loc, MultipleFlowAttributes) :: errors, info)
          else
            (errors, { info with flow = Some OptInStrict })
        in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@flow") :: (_, "strict-local") :: xs ->
        let acc =
          if info.flow <> None then
            ((loc, MultipleFlowAttributes) :: errors, info)
          else
            (errors, { info with flow = Some OptInStrictLocal })
        in
        parse_attributes ~file_options ~filename acc xs
      | (flow_loc, "@flow") :: (next_loc, next) :: xs
        when Loc.lines_intersect flow_loc next_loc && (not @@ Str.string_match pragma_rx next 0) ->
        let (errors, info) =
          if info.flow <> None then
            ((flow_loc, MultipleFlowAttributes) :: errors, info)
          else
            (errors, { info with flow = Some OptIn })
        in
        let errors = (next_loc, InvalidFlowMode next) :: errors in
        parse_attributes ~file_options ~filename (errors, info) xs
      | (loc, "@flow") :: xs ->
        let acc =
          if info.flow <> None then
            ((loc, MultipleFlowAttributes) :: errors, info)
          else
            (errors, { info with flow = Some OptIn })
        in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@noflow") :: xs ->
        let acc =
          if info.flow <> None then
            ((loc, MultipleFlowAttributes) :: errors, info)
          else
            (errors, { info with flow = Some OptOut })
        in
        parse_attributes ~file_options ~filename acc xs
      | (_, "@preventMunge") :: xs ->
        (* dupes are ok since they can only be truthy *)
        parse_attributes ~file_options ~filename (errors, { info with preventMunge = true }) xs
      | [(jsx_loc, "@jsx")] -> ((jsx_loc, InvalidJSXAttribute None) :: errors, info)
      | (jsx_loc, "@jsx") :: (expr_loc, expr) :: xs ->
        let acc =
          if info.jsx <> None then
            ((jsx_loc, MultipleJSXAttributes) :: errors, info)
          else
            (* The point of the padding is to make the parsed code line up
             * with the comment in the original source *)
            let padding =
              String.make Loc.(expr_loc.start.line - 1) '\n'
              ^ String.make Loc.(expr_loc.start.column) ' '
            in
            try
              let (jsx_expr, _) =
                Parser_flow.jsx_pragma_expression (padding ^ expr) expr_loc.Loc.source
              in
              (errors, { info with jsx = Some (expr, jsx_expr) })
            with
            | Parse_error.Error ((_, e), _) ->
              let e = Some (Parse_error.PP.error e) in
              ((expr_loc, InvalidJSXAttribute e) :: errors, info)
        in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@jsxRuntime") :: (_, "classic") :: xs ->
        let acc =
          if info.jsxRuntime <> None then
            ((loc, MultipleJSXRuntimeAttributes) :: errors, info)
          else
            (errors, { info with jsxRuntime = Some JsxRuntimePragmaClassic })
        in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@jsxRuntime") :: (_, "automatic") :: xs ->
        let acc =
          if info.jsxRuntime <> None then
            ((loc, MultipleJSXRuntimeAttributes) :: errors, info)
          else
            (errors, { info with jsxRuntime = Some JsxRuntimePragmaAutomatic })
        in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@jsxRuntime") :: _ :: xs ->
        let acc = ((loc, InvalidJSXRuntimeAttribute) :: errors, info) in
        parse_attributes ~file_options ~filename acc xs
      | (loc, "@supportsPlatform") :: (_, platform) :: xs when Files.multi_platform file_options ->
        let acc =
          if
            filename
            |> File_key.to_string
            |> Files.platform_specific_extensions_and_indices_opt ~options:file_options
            |> Option.is_some
          then
            ((loc, DisallowedSupportsPlatform) :: errors, info)
          else if
            Base.List.mem
              ~equal:String.equal
              (Files.multi_platform_extensions file_options)
              ("." ^ platform)
          then
            let existing_platforms = Base.Option.value info.supportsPlatform ~default:[] in
            let platforms =
              if Base.List.mem ~equal:String.equal existing_platforms platform then
                existing_platforms
              else
                platform :: existing_platforms
            in
            (errors, { info with supportsPlatform = Some platforms })
          else
            ((loc, InvalidSupportsPlatform platform) :: errors, info)
        in
        parse_attributes ~file_options ~filename acc xs
      | _ :: xs -> parse_attributes ~file_options ~filename (errors, info) xs
      | [] -> (errors, info)
    in
    let calc_end start s =
      Str.full_split lines_rx s
      |> List.fold_left
           Loc.(
             fun _end elem ->
               match elem with
               | Str.Delim delim ->
                 let line_incr =
                   if delim = "\r" then
                     0
                   else
                     1
                 in
                 let column = 0 in
                 let line = _end.line + line_incr in
                 { column; line }
               | Str.Text text ->
                 let length = String.length text in
                 let column = _end.column + length in
                 { _end with column }
           )
           start
    in
    let split loc s =
      (* Need to add 2 characters for the start of the comment *)
      let start = Loc.{ loc.start with column = loc.start.column + 2 } in
      Str.full_split attributes_rx s
      |> List.fold_left
           (fun (start, attributes) elem ->
             match elem with
             | Str.Delim s -> (calc_end start s, attributes)
             | Str.Text s ->
               let _end = calc_end start s in
               (_end, Loc.({ loc with start; _end }, s) :: attributes))
           (start, [])
      |> snd
      |> List.rev
    in
    let string_of_comment = function
      | (loc, { Ast.Comment.text; _ }) -> (loc, text)
    in
    let map_n =
      let rec helper f remaining acc = function
        | [] -> List.rev acc
        | hd :: rest ->
          if remaining <= 0 then
            List.rev acc
          else
            helper f (remaining - 1) (f hd :: acc) rest
      in
      (fun f n lst -> helper f n [] lst)
    in
    fun ~max_tokens ~file_options filename content ->
      (* Consume tokens in the file until we get a comment. This is a hack to
       * support Nuclide, which needs 'use babel' as the first token due to
       * constraints with Atom (see https://github.com/atom/atom/issues/8416 for
       * more context). At some point this should change back to consuming only
       * the first token. *)
      let lb =
        try Sedlexing.Utf8.from_string content with
        | Sedlexing.MalFormed ->
          Hh_logger.warn "File %s is malformed" (File_key.to_string filename);
          Sedlexing.Utf8.from_string ""
      in
      let env = Lex_env.new_lex_env (Some filename) lb ~enable_types_in_comments:false in
      let rec get_first_comment_contents ?(i = 0) env =
        if i < max_tokens then
          let (env, lexer_result) = Flow_lexer.token env in
          match Lex_result.comments lexer_result with
          | [] ->
            Token.(
              (*
               * Stop looking for docblocks if we see any tokens other than a
               * string or a semicolon (`"use babel";` or `"use strict";`) or
               * an interpreter directive (`#!/bin/env node`)
               *)
              (match Lex_result.token lexer_result with
              | T_STRING _
              | T_SEMICOLON
              | T_INTERPRETER _ ->
                get_first_comment_contents ~i:(i + 1) env
              | _ -> None)
            )
          | comments -> Some (map_n string_of_comment (max_tokens - i) comments)
        else
          None
      in
      match get_first_comment_contents env with
      | Some comments ->
        List.fold_left
          (fun acc (loc, s) -> parse_attributes ~file_options ~filename acc (split loc s))
          ([], default_info)
          comments
      | None -> ([], default_info)
  )

let parse_docblock ~max_tokens ~file_options file content : docblock_error list * Docblock.t =
  match file with
  | File_key.ResourceFile _
  | File_key.JsonFile _ ->
    ([], Docblock.default_info)
  | _ -> extract_docblock ~max_tokens ~file_options file content

File: rust_port/crates/flow_parsing/src/docblock_parser.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::VecDeque;
use std::sync::Arc;

use dupe::Dupe;
use flow_common::docblock::Docblock;
use flow_common::docblock::FlowMode;
use flow_common::docblock::JsxPragma;
use flow_common::docblock::JsxRuntimePragma;
use flow_common::files::FileOptions;
use flow_common::files::platform_specific_extensions_and_indices_opt;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::file_key::FileKey;
use flow_parser::file_key::FileKeyInner;
use flow_parser::flow_lexer;
use flow_parser::lex_env::LexEnv;
use flow_parser::loc::Loc;
use flow_parser::loc::Position;
use flow_parser::logos_tokens::MainToken;
use flow_parser::parse_jsx_pragma_expression;
use flow_parser::token::TokenKind;
use lazy_static::lazy_static;
use logos::Logos;
use regex::Regex;

// Avoid lexing unbounded in perverse cases
pub const DOCBLOCK_MAX_TOKENS: usize = 10;

lazy_static! {
    static ref ATTRIBUTES_RX: Regex = Regex::new(r"[ \t\r\n\\*/]+").unwrap();
    static ref LINES_RX: Regex = Regex::new(r"(\r\n|\n|\r)").unwrap();
}

#[derive(Debug, Clone, PartialEq)]
pub enum DocblockErrorKind {
    MultipleFlowAttributes,
    InvalidFlowMode(String),
    MultipleJSXAttributes,
    InvalidJSXAttribute(Option<String>),
    MultipleJSXRuntimeAttributes,
    InvalidJSXRuntimeAttribute,
    InvalidSupportsPlatform(String),
    DisallowedSupportsPlatform,
}

#[derive(Debug, Clone, PartialEq)]
pub struct DocblockError {
    pub loc: Loc,
    pub kind: DocblockErrorKind,
}

struct Attribute {
    loc: Loc,
    text: String,
}

fn extract_docblock(
    max_tokens: usize,
    file_options: &FileOptions,
    file_key: &FileKey,
    content: &str,
) -> (Vec<DocblockError>, Docblock) {
    // walks a list of words, returns a list of errors and the extracted info.
    // if @flow or @providesModule is found more than once, the first one is used
    // and an error is returned.
    fn parse_attributes(
        file_options: &FileOptions,
        file_key: &FileKey,
        info: &mut Docblock,
        errors: &mut Vec<DocblockError>,
        mut attributes: VecDeque<Attribute>,
    ) {
        while let Some(attr) = attributes.pop_front() {
            match attr.text.as_str() {
                "@flow" => {
                    if let Some(next_attr) = attributes.pop_front() {
                        if next_attr.text == "strict" {
                            if info.flow.is_some() {
                                errors.push(DocblockError {
                                    loc: attr.loc,
                                    kind: DocblockErrorKind::MultipleFlowAttributes,
                                });
                            } else {
                                info.flow = Some(FlowMode::OptInStrict);
                            }
                            continue;
                        } else if next_attr.text == "strict-local" {
                            if info.flow.is_some() {
                                errors.push(DocblockError {
                                    loc: attr.loc,
                                    kind: DocblockErrorKind::MultipleFlowAttributes,
                                });
                            } else {
                                info.flow = Some(FlowMode::OptInStrictLocal);
                            }
                            continue;
                        } else if attr.loc.lines_intersect(&next_attr.loc)
                            && !next_attr.text.starts_with('@')
                        {
                            if info.flow.is_some() {
                                errors.push(DocblockError {
                                    loc: attr.loc.dupe(),
                                    kind: DocblockErrorKind::MultipleFlowAttributes,
                                });
                            } else {
                                info.flow = Some(FlowMode::OptIn);
                            }
                            errors.push(DocblockError {
                                loc: next_attr.loc,
                                kind: DocblockErrorKind::InvalidFlowMode(next_attr.text.clone()),
                            });
                            continue;
                        } else {
                            // Put back the next attribute, we didn't consume it
                            attributes.push_front(next_attr);
                        }
                    }

                    if info.flow.is_some() {
                        errors.push(DocblockError {
                            loc: attr.loc,
                            kind: DocblockErrorKind::MultipleFlowAttributes,
                        });
                    } else {
                        info.flow = Some(FlowMode::OptIn);
                    }
                }
                "@noflow" => {
                    if info.flow.is_some() {
                        errors.push(DocblockError {
                            loc: attr.loc.dupe(),
                            kind: DocblockErrorKind::MultipleFlowAttributes,
                        });
                    } else {
                        info.flow = Some(FlowMode::OptOut);
                    }
                }
                "@preventMunge" => {
                    info.prevent_munge = true;
                }
                "@jsx" => {
                    if let Some(expr_attr) = attributes.pop_front() {
                        if info.jsx.is_some() {
                            errors.push(DocblockError {
                                loc: attr.loc.dupe(),
                                kind: DocblockErrorKind::MultipleJSXAttributes,
                            });
                        } else {
                            // The point of the padding is to make the parsed code line up
                            // with the comment in the original source
                            let padding = "\n".repeat((expr_attr.loc.start.line - 1) as usize)
                                + &" ".repeat(expr_attr.loc.start.column as usize);
                            let mut padded_expr = padding;
                            padded_expr.push_str(&expr_attr.text);

                            match parse_jsx_pragma_expression::<()>(
                                expr_attr.loc.source.dupe(),
                                &padded_expr,
                            ) {
                                Ok((jsx_expr, _)) => {
                                    info.jsx = Some(Arc::new(JsxPragma {
                                        raw: expr_attr.text,
                                        expression: jsx_expr,
                                    }));
                                }
                                Err(parse_errors) => {
                                    let first_error =
                                        parse_errors.first().map(|(_, e)| e.to_string());
                                    errors.push(DocblockError {
                                        loc: expr_attr.loc,
                                        kind: DocblockErrorKind::InvalidJSXAttribute(first_error),
                                    });
                                }
                            }
                        }
                    } else {
                        errors.push(DocblockError {
                            loc: attr.loc,
                            kind: DocblockErrorKind::InvalidJSXAttribute(None),
                        });
                    }
                }
                "@jsxRuntime" => {
                    if let Some(runtime_attr) = attributes.pop_front() {
                        if runtime_attr.text == "classic" {
                            if info.jsx_runtime.is_some() {
                                errors.push(DocblockError {
                                    loc: attr.loc,
                                    kind: DocblockErrorKind::MultipleJSXRuntimeAttributes,
                                });
                            } else {
                                info.jsx_runtime = Some(JsxRuntimePragma::Classic);
                            }
                        } else if runtime_attr.text == "automatic" {
                            if info.jsx_runtime.is_some() {
                                errors.push(DocblockError {
                                    loc: attr.loc,
                                    kind: DocblockErrorKind::MultipleJSXRuntimeAttributes,
                                });
                            } else {
                                info.jsx_runtime = Some(JsxRuntimePragma::Automatic);
                            }
                        } else {
                            errors.push(DocblockError {
                                loc: attr.loc,
                                kind: DocblockErrorKind::InvalidJSXRuntimeAttribute,
                            });
                        }
                    } else {
                        errors.push(DocblockError {
                            loc: attr.loc,
                            kind: DocblockErrorKind::InvalidJSXRuntimeAttribute,
                        });
                    }
                }
                "@supportsPlatform" => {
                    if let Some(platform_attr) = attributes.pop_front() {
                        if file_options.multi_platform {
                            let platform = &platform_attr.text;

                            if platform_specific_extensions_and_indices_opt(
                                file_options,
                                file_key.as_str(),
                            )
                            .is_some()
                            {
                                errors.push(DocblockError {
                                    loc: attr.loc.dupe(),
                                    kind: DocblockErrorKind::DisallowedSupportsPlatform,
                                });
                            } else if file_options
                                .multi_platform_extensions
                                .iter()
                                .any(|ext| ext.as_str() == format!(".{}", platform))
                            {
                                let platform_smol = FlowSmolStr::from(platform.as_str());
                                let mut existing_platforms =
                                    info.supports_platform.clone().unwrap_or_else(Vec::new);
                                if !existing_platforms.contains(&platform_smol) {
                                    existing_platforms.push(platform_smol);
                                }
                                info.supports_platform = Some(existing_platforms);
                            } else {
                                errors.push(DocblockError {
                                    loc: attr.loc.dupe(),
                                    kind: DocblockErrorKind::InvalidSupportsPlatform(
                                        platform.clone(),
                                    ),
                                });
                            }
                        }
                    }
                }
                _ => {}
            }
        }
    }

    fn calc_end(start: Position, s: &str) -> Position {
        let mut end = start;

        let mut last_pos = 0;
        for cap in LINES_RX.find_iter(s) {
            let text_len = (cap.start() - last_pos) as i32;
            end.column += text_len;
            let delim = cap.as_str();
            let line_incr = if delim == "\r" { 0 } else { 1 };
            end = Position {
                line: end.line + line_incr,
                column: 0,
            };

            last_pos = cap.end();
        }

        let remaining_len = (s.len() - last_pos) as i32;
        end.column += remaining_len;

        end
    }

    fn split(loc: Loc, text: &str) -> VecDeque<Attribute> {
        // Need to add 2 characters for the start of the comment
        let mut start = Position {
            line: loc.start.line,
            column: loc.start.column + 2,
        };
        let mut attributes = VecDeque::new();
        let mut last_pos = 0;

        for mat in ATTRIBUTES_RX.find_iter(text) {
            if mat.start() > last_pos {
                let part = &text[last_pos..mat.start()];
                let end = calc_end(start, part);
                attributes.push_back(Attribute {
                    loc: Loc {
                        source: loc.source.dupe(),
                        start,
                        end,
                    },
                    text: part.to_string(),
                });
                start = end;
            }
            let delim = mat.as_str();
            start = calc_end(start, delim);

            last_pos = mat.end();
        }

        if last_pos < text.len() {
            let part = &text[last_pos..];
            let end = calc_end(start, part);
            attributes.push_back(Attribute {
                loc: Loc {
                    source: loc.source.dupe(),
                    start,
                    end,
                },
                text: part.to_string(),
            });
        }

        attributes
    }

    fn string_of_comment(comment: &flow_parser::ast::Comment<Loc>) -> (Loc, Arc<str>) {
        (comment.loc.dupe(), comment.text.dupe())
    }

    // Consume tokens in the file until we get a comment. This is a hack to
    // support Nuclide, which needs 'use babel' as the first token due to
    // constraints with Atom (see https://github.com/atom/atom/issues/8416 for
    // more context). At some point this should change back to consuming only
    // the first token.
    let mut lexer = MainToken::lexer(content);
    let mut env = LexEnv::new(Some(file_key.dupe()), false);

    fn get_first_comment_contents(
        lexer: &mut logos::Lexer<MainToken>,
        env: &mut LexEnv,
        max_tokens: usize,
        i: usize,
    ) -> Option<Vec<(Loc, Arc<str>)>> {
        if i < max_tokens {
            let lex_result = flow_lexer::token(lexer, env);

            if !lex_result.comments.is_empty() {
                Some(
                    lex_result
                        .comments
                        .iter()
                        .take(max_tokens - i)
                        .map(string_of_comment)
                        .collect(),
                )
            } else {
                // Stop looking for docblocks if we see any tokens other than a
                // string or a semicolon (`"use babel";` or `"use strict";`) or
                // an interpreter directive (`#!/bin/env node`)
                match lex_result.token_kind {
                    TokenKind::TString(..)
                    | TokenKind::TSemicolon
                    | TokenKind::TInterpreter(..) => {
                        get_first_comment_contents(lexer, env, max_tokens, i + 1)
                    }
                    _ => None,
                }
            }
        } else {
            None
        }
    }

    match get_first_comment_contents(&mut lexer, &mut env, max_tokens, 0) {
        Some(comments) => {
            let mut errors = Vec::new();
            let mut info = Docblock::default();

            for (loc, text) in comments {
                let attrs = split(loc, &text);
                parse_attributes(file_options, file_key, &mut info, &mut errors, attrs);
            }

            (errors, info)
        }
        None => (Vec::new(), Docblock::default()),
    }
}

pub fn parse_docblock(
    max_tokens: usize,
    file_options: &FileOptions,
    file_key: &FileKey,
    content: &str,
) -> (Vec<DocblockError>, Docblock) {
    match file_key.inner() {
        FileKeyInner::ResourceFile(_) | FileKeyInner::JsonFile(_) => {
            (Vec::new(), Docblock::default())
        }
        _ => extract_docblock(max_tokens, file_options, file_key, content),
    }
}
